src/functional/ffront/dialect_parser.py

```diff
@@ -131,8 +135,37 @@ def apply(
             #  does not require passing the information on every invocation.
             if not err.filename:
                 err.filename = filename
+
+            # ensure line numbers are relative to the file (respects `starting_line`)
+            if err.lineno:
+                err.lineno = err.lineno + line_offset
+            if err.end_lineno:
+                err.end_lineno = err.end_lineno + line_offset
+
+            # if offsets are provided, but not line numbers, things get weird…
+            assert err.lineno or not err.offset
+            assert err.end_lineno or not err.end_offset
+            # if the ends are provided, but not the starts, things get weird…
+            assert err.lineno or not err.end_lineno
+            assert err.offset or not err.end_offset
+
             if not err.text:
-                err.text = source
+
+                if err.lineno:
+
+                    source_lineno = err.lineno - starting_line
+                    source_end_lineno = (
+                        (err.end_lineno - starting_line) if err.end_lineno else source_lineno
+                    )
+                    err.text = "\n".join(source.split("\n")[source_lineno : source_end_lineno + 1])
+
+                else:
+                    err.text = source
+
+                    # `err.offset` determines if carrets (`^^^^`) are printed below `err.text`.
```

Comment:
Is this related to the same explanation in lines 56-58 of  `test_error_line_number.py`? Could you add a similar comment here to make it simpler to understand?